File: pwa/src/port-catalog.test.ts
/**
 * The grouping rules, pinned against the phone's.
 *
 * Two clients now derive the same six groups from the same three inputs, in two
 * languages that cannot import each other. What holds them together is this file
 * and `ios/Tests/PortCatalogTests.swift` asking the same questions — so the table
 * in `port-catalog.ts` is not documentation of an intention, it is a thing that
 * fails when somebody changes it on one side only.
 *
 * The rule these tests exist to defend, above all the others: **no group is ever
 * derived from a port number**. `dev-ports.ts` refuses to guess which framework
 * is behind a port and neither client may do the guessing on its behalf, because
 * a port number is a number a person chose. Two of the cases below check that
 * directly, with 3000 and 5173 landing wherever their *process* puts them.
 */

import { describe, expect, it } from 'vitest'
import { BRAND } from '../../src/shared/brand'
import {
  PORT_CATEGORIES,
  categoryTitle,
  directAppPorts,
  foldedByDefault,
  isOwnProcess,
  isWebRuntime,
  portRowDetail,
  portRowTitle,
  secondAction,
  sections,
} from './port-catalog'
import type { DevServerReport, LocalPort } from './protocol-client'

function port(number: number, process: string, guessed = false): LocalPort {
  return { port: number, process, guessed }
}

function dev(folder: string, status: DevServerReport['status'], extra: Partial<DevServerReport> = {}): DevServerReport {
  return { folder, status, ...extra }
}

describe('the six groups', () => {
  it('goes from why you opened the screen to the noise, and folds the noise', () => {
    // The declaration order is the draw order, so it is the thing to assert.
    expect([...PORT_CATEGORIES]).toEqual(['named', 'devServer', 'web', 'app', 'other', 'unnamed'])
    expect(PORT_CATEGORIES.filter(foldedByDefault)).toEqual(['app', 'other', 'unnamed'])
  })

  it('calls the app’s own group by the product’s name', () => {
    // Not "this app". A browser paired with three machines is answering *which
    // program on that machine* is holding the port — and the name is read from
    // the one place it lives.
    expect(categoryTitle('app')).toBe(BRAND.name)
    expect(categoryTitle('unnamed')).toBe('Unidentified')
  })

  it('never sorts a port into a group by its number', () => {
    const grouped = sections({
      ports: [port(3000, 'wslrelay'), port(5173, 'node'), port(8080, 'unknown', true)],
      devServers: [],
    })
    // 3000 is the number everybody associates with a dev server and it is held by
    // a relay here, so it is "other". 5173 is Vite's number and is `node` here, so
    // it is a web server — on the process, not on the number.
    expect(grouped.map((section) => section.category)).toEqual(['web', 'other', 'unnamed'])
    expect(grouped[0].rows[0].port).toBe(5173)
    expect(grouped[1].rows[0].port).toBe(3000)
    expect(grouped[2].rows[0].port).toBe(8080)
  })

  it('recognises the runtimes as prefixes, because the scanners spell them differently', () => {
    // `python3` from lsof, `node.exe` already stripped by tasklist but sometimes
    // carrying a suffix. Matching exactly would put half of them in "Other".
    expect(isWebRuntime('python3')).toBe(true)
    expect(isWebRuntime('NODE')).toBe(true)
    expect(isWebRuntime('nodemon')).toBe(true)
    expect(isWebRuntime('wslrelay')).toBe(false)
  })

  it('knows the product’s own process under both of its spellings', () => {
    expect(isOwnProcess(BRAND.name)).toBe(true)
    expect(isOwnProcess(BRAND.name.replace(' ', ''))).toBe(true)
    expect(isOwnProcess(BRAND.id)).toBe(true)
    expect(isOwnProcess('node')).toBe(false)
  })

  it('puts the socket this page is connected on in the app’s group, before the runtime name', () => {
    /*
     * The load-bearing ordering. A desktop running headless *is* a node process,
     * so on a direct pairing the client's own control socket would otherwise be
     * offered under "Web servers" — a row describing the thing that drew it.
     */
    const grouped = sections({
      ports: [port(8443, 'node')],
      devServers: [],
      appPorts: [8443],
    })
    expect(grouped).toHaveLength(1)
    expect(grouped[0].category).toBe('app')
  })
})

describe('naming a port', () => {
  it('lifts it to the top group whatever it was derived into', () => {
    const grouped = sections({
      ports: [port(2019, 'wslrelay'), port(2222, 'wslrelay')],
      devServers: [],
      names: { 2222: 'WSL relay — ssh' },
    })
    expect(grouped.map((section) => section.category)).toEqual(['named', 'other'])
    expect(grouped[0].rows[0].port).toBe(2222)
  })

  it('leads with the name and moves the address to the second line', () => {
    const named = sections({ ports: [port(3210, 'node')], devServers: [], names: { 3210: 'Client billing' } })
    expect(portRowTitle(named[0].rows[0])).toBe('Client billing')
    expect(portRowDetail(named[0].rows[0])).toBe('localhost:3210 · node')

    const plain = sections({ ports: [port(3210, 'node')], devServers: [] })
    expect(portRowTitle(plain[0].rows[0])).toBe('localhost:3210')
    expect(portRowDetail(plain[0].rows[0])).toBe('node')
  })

  it('says nothing about a process the machine could not name', () => {
    // `guessed` means the port answers and nothing could name its owner. Printing
    // the literal string `unknown` would be a field dump; omitting the line is the
    // honest reading.
    const grouped = sections({ ports: [port(9000, 'unknown', true)], devServers: [] })
    expect(portRowDetail(grouped[0].rows[0])).toBeNull()
  })
})

describe('one row per server, never two', () => {
  it('joins a ready dev server to the port it is proven to be serving on', () => {
    const grouped = sections({
      ports: [port(5173, 'node'), port(2019, 'wslrelay')],
      devServers: [dev('/w/site', 'ready', { port: 5173, sessionId: 's1', url: 'http://localhost:5173' })],
    })
    // One dev-server row carrying both halves, and no separate `5173 · node`.
    const rows = grouped.flatMap((section) => section.rows)
    expect(rows.filter((row) => row.port === 5173)).toHaveLength(1)
    const joined = rows.find((row) => row.port === 5173)
    expect(joined?.dev?.folder).toBe('/w/site')
    expect(joined?.entry?.process).toBe('node')
    expect(joined?.category).toBe('devServer')
  })

  it('does not let a starting or failed report claim a port', () => {
    /*
     * Only a `ready` report has a proven port. A `starting` one has no port field
     * and a `failed` one must never carry the address of the server that died — so
     * a port row that happens to share the number is a *different* thing and stays
     * on the list in its own right.
     */
    const grouped = sections({
      ports: [port(5173, 'node')],
      devServers: [dev('/w/site', 'failed', { port: 5173, message: 'exited 1' })],
    })
    const rows = grouped.flatMap((section) => section.rows)
    expect(rows).toHaveLength(2)
    expect(rows.some((row) => row.dev !== null && row.entry === null)).toBe(true)
    expect(rows.some((row) => row.dev === null && row.entry !== null)).toBe(true)
  })

  it('gives a named dev server the top group too', () => {
    const grouped = sections({
      ports: [],
      devServers: [dev('/w/site', 'ready', { port: 4321, sessionId: 's1' })],
      names: { 4321: 'The shop' },
    })
    expect(grouped[0].category).toBe('named')
    expect(grouped[0].rows[0].name).toBe('The shop')
  })

  it('never draws a folder with no dev script', () => {
    // It means "there is nothing to press, and there never will be for this
    // folder". A row for it would be a button whose only outcome is a refusal.
    const grouped = sections({ ports: [], devServers: [dev('/w/notes', 'no-dev-script')] })
    expect(grouped).toEqual([])
  })

  it('keeps the order the two lists arrived in', () => {
    // The desktop ranks its ports most-likely-to-be-a-dev-server first. Re-sorting
    // here would throw away the only ordering anybody has an opinion about.
    const grouped = sections({
      ports: [port(5173, 'node'), port(3000, 'bun'), port(8000, 'python3')],
      devServers: [],
    })
    expect(grouped[0].rows.map((row) => row.port)).toEqual([5173, 3000, 8000])
  })
})

describe('what the row’s second control does', () => {
  it('answers all five dev-server states and the plain port', () => {
    const idle = sections({ ports: [], devServers: [dev('/w/a', 'idle')] })[0].rows[0]
    expect(secondAction(idle)).toEqual({ kind: 'start', folder: '/w/a' })

    const failed = sections({ ports: [], devServers: [dev('/w/a', 'failed')] })[0].rows[0]
    expect(secondAction(failed)).toEqual({ kind: 'retry', folder: '/w/a' })

    const starting = sections({ ports: [], devServers: [dev('/w/a', 'starting', { sessionId: 's7' })] })[0].rows[0]
    expect(secondAction(starting)).toEqual({ kind: 'openSession', id: 's7' })

    const ready = sections({
      ports: [],
      devServers: [dev('/w/a', 'ready', { port: 3000, sessionId: 's8' })],
    })[0].rows[0]
    expect(secondAction(ready)).toEqual({ kind: 'openSession', id: 's8' })

    const plain = sections({ ports: [port(2019, 'wslrelay')], devServers: [] })[0].rows[0]
    expect(secondAction(plain)).toEqual({ kind: 'copyAddress', port: 2019 })
  })

  it('offers nothing rather than a control with nowhere to go', () => {
    // The protocol says a running dev server always carries a session. Drawing a
    // button that would have nothing to open is worse than drawing none, so the
    // impossible case is handled rather than forced.
    const ready = sections({ ports: [], devServers: [dev('/w/a', 'ready', { port: 3000 })] })[0].rows[0]
    expect(secondAction(ready)).toEqual({ kind: 'none' })
  })
})

describe('which port belongs to this product', () => {
  it('takes the page’s own port for a direct pairing', () => {
    expect(directAppPorts({ protocol: 'https:', port: '8443' })).toEqual([8443])
    // No explicit port means the scheme's default, which is a fact about the URL
    // rather than an assumption about the machine.
    expect(directAppPorts({ protocol: 'https:', port: '' })).toEqual([443])
    expect(directAppPorts({ protocol: 'http:', port: '' })).toEqual([80])
    expect(directAppPorts({ protocol: 'file:', port: '' })).toEqual([])
  })

  it('claims nothing on a relay pairing, because nothing on this side knows', () => {
    /*
     * The browser dials the relay and the desktop dials out to meet it, so no
     * frame on this side carries the local port the desktop bound. Falling back to
     * the product's default port number would be a guess about somebody's
     * configuration — the exact thing this whole file is built on not making — so
     * the caller passes nothing and the app's group is simply empty.
     */
    const grouped = sections({ ports: [port(8443, 'node')], devServers: [], appPorts: [] })
    expect(grouped[0].category).toBe('web')
  })
})

File: android/app/src/test/java/dev/terminaldeck/android/ports/PortCatalogTest.kt
package dev.terminaldeck.android.ports

import dev.terminaldeck.android.protocol.DevServerReport
import dev.terminaldeck.android.protocol.DevServerStatus
import dev.terminaldeck.android.protocol.LocalPort
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The grouping, which is what turns a wall of ports into a screen.
 *
 * Every rule is pinned here rather than checked by looking at a phone, because the input that
 * produces the interesting cases — a WSL box with four `wslrelay` ports and a dev server on 5173 — is
 * not something anybody has to hand.
 */
class PortCatalogTest {

    private fun port(number: Int, process: String = "node", guessed: Boolean = false) =
        LocalPort(port = number, process = process, guessed = guessed)

    private fun dev(
        folder: String,
        status: DevServerStatus,
        port: Int? = null,
        session: String? = null,
    ) = DevServerReport(
        folder = folder,
        status = status,
        script = "dev",
        command = "pnpm run dev",
        sessionId = session,
        port = port,
        url = port?.let { "http://localhost:$it" },
    )

    private fun categoriesOf(sections: List<LocalhostSection>) = sections.map { it.category }

    @Test
    fun `a runtime that usually serves a page lands in web servers`() {
        val sections = PortCatalog.sections(listOf(port(5173, "node")), emptyList())
        assertEquals(listOf(PortCategory.Web), categoriesOf(sections))
    }

    @Test
    fun `a runtime is matched as a prefix, because two scanners spell them differently`() {
        assertTrue(PortCatalog.isWebRuntime("python3"))
        assertTrue(PortCatalog.isWebRuntime("Node"))
        assertTrue(PortCatalog.isWebRuntime("node22"))
        assertTrue(!PortCatalog.isWebRuntime("wslrelay"))
    }

    @Test
    fun `an unowned port is its own group, not part of the dull pile`() {
        val sections = PortCatalog.sections(listOf(port(2019, "unknown", guessed = true)), emptyList())
        // "we could not name this" and "this is named and dull" are different facts, and only one of
        // them might be worth a second look.
        assertEquals(listOf(PortCategory.Unnamed), categoriesOf(sections))
    }

    @Test
    fun `a named process that is nothing else is other services`() {
        val sections = PortCatalog.sections(listOf(port(6666, "AgentService")), emptyList())
        assertEquals(listOf(PortCategory.Other), categoriesOf(sections))
    }

    @Test
    fun `this product's own listener is not offered as somebody's dev server`() {
        // A desktop running headless is a `node` process, so the app's own port has to be recognised
        // before the runtime name is looked at — otherwise the phone offers itself its own socket.
        val sections = PortCatalog.sections(listOf(port(7777, "node")), emptyList(), appPorts = setOf(7777))
        assertEquals(listOf(PortCategory.App), categoriesOf(sections))

        assertTrue(PortCatalog.isOwnProcess("Terminal Deck"))
        assertTrue(PortCatalog.isOwnProcess("TerminalDeck"))
        assertTrue(PortCatalog.isOwnProcess("terminaldeck"))
        assertTrue(!PortCatalog.isOwnProcess("terminal"))
    }

    @Test
    fun `naming a port promotes it out of whatever pile it was in`() {
        val sections = PortCatalog.sections(
            ports = listOf(port(2019, "wslrelay"), port(5173, "node")),
            devServers = emptyList(),
            names = mapOf(2019 to "The relay"),
        )
        // One gesture with one meaning — *this one matters and here is why* — rather than a second
        // pin control that could get out of step with the name.
        assertEquals(listOf(PortCategory.Named, PortCategory.Web), categoriesOf(sections))
        assertEquals("The relay", sections.first().rows.single().name)
    }

    @Test
    fun `a ready dev server claims its port, so there is one row and not two`() {
        val sections = PortCatalog.sections(
            ports = listOf(port(5173, "node")),
            devServers = listOf(dev("/w/app", DevServerStatus.Ready, port = 5173, session = "s1")),
        )
        assertEquals(listOf(PortCategory.DevServer), categoriesOf(sections))
        val row = sections.single().rows.single()
        // The merged row carries both the project and the address.
        assertEquals("/w/app", row.dev?.folder)
        assertEquals(5173, row.entry?.port)
        assertEquals(5173, row.port)
    }

    @Test
    fun `a starting dev server has no port, so nothing is claimed and nothing is dropped`() {
        val sections = PortCatalog.sections(
            ports = listOf(port(5173, "node")),
            devServers = listOf(dev("/w/app", DevServerStatus.Starting, session = "s1")),
        )
        assertEquals(listOf(PortCategory.DevServer, PortCategory.Web), categoriesOf(sections))
        assertNull(sections.first().rows.single().port)
    }

    @Test
    fun `a failed dev server never carries the address of the server that died`() {
        // The report may not even mention a port; if it did, this must not join it to a live row.
        val sections = PortCatalog.sections(
            ports = listOf(port(5173, "node")),
            devServers = listOf(dev("/w/app", DevServerStatus.Failed, port = 5173)),
        )
        val devRow = sections.first { it.category == PortCategory.DevServer }.rows.single()
        assertNull(devRow.entry)
        // And the listening port is still its own row, because nothing claimed it.
        assertTrue(sections.any { it.category == PortCategory.Web })
    }

    @Test
    fun `no-dev-script is never a row`() {
        val sections = PortCatalog.sections(
            ports = emptyList(),
            devServers = listOf(dev("/w/app", DevServerStatus.NoDevScript)),
        )
        assertEquals(emptyList<LocalhostSection>(), sections)
    }

    @Test
    fun `the sections come out in the order they are drawn in`() {
        val sections = PortCatalog.sections(
            ports = listOf(
                port(2019, "unknown", guessed = true),
                port(6666, "AgentService"),
                port(7777, "node"),
                port(5173, "node"),
                port(3000, "node"),
            ),
            devServers = listOf(dev("/w/app", DevServerStatus.Idle)),
            appPorts = setOf(7777),
            names = mapOf(3000 to "The API"),
        )
        assertEquals(
            listOf(
                PortCategory.Named,
                PortCategory.DevServer,
                PortCategory.Web,
                PortCategory.App,
                PortCategory.Other,
                PortCategory.Unnamed,
            ),
            categoriesOf(sections),
        )
    }

    @Test
    fun `order within a section is the order the machine sent`() {
        val sections = PortCatalog.sections(
            ports = listOf(port(5173, "node"), port(3000, "node"), port(8080, "node")),
            devServers = emptyList(),
        )
        // The desktop ranks its ports most-likely-to-be-a-dev-server first, and re-sorting here would
        // throw away the only ordering anybody has an opinion about.
        assertEquals(listOf(5173, 3000, 8080), sections.single().rows.map { it.port })
    }

    @Test
    fun `the three noisy groups start folded and the three useful ones do not`() {
        assertEquals(false, PortCategory.Named.foldedByDefault)
        assertEquals(false, PortCategory.DevServer.foldedByDefault)
        assertEquals(false, PortCategory.Web.foldedByDefault)
        assertEquals(true, PortCategory.App.foldedByDefault)
        assertEquals(true, PortCategory.Other.foldedByDefault)
        assertEquals(true, PortCategory.Unnamed.foldedByDefault)
    }

    @Test
    fun `a row's id is stable across a port changing under a dev server`() {
        val first = PortCatalog.sections(emptyList(), listOf(dev("/w/app", DevServerStatus.Ready, port = 5173)))
        val second = PortCatalog.sections(emptyList(), listOf(dev("/w/app", DevServerStatus.Ready, port = 5174)))
        // Vite takes 5174 when 5173 is busy. Keying the row on the port would re-animate it.
        assertEquals(first.single().rows.single().id, second.single().rows.single().id)
    }

    @Test
    fun `the second action is the one the state deserves`() {
        fun action(status: DevServerStatus, session: String? = "s1") =
            PortCatalog.secondAction(
                LocalhostRow(null, dev("/w/app", status, session = session), null, PortCategory.DevServer)
            )

        assertEquals(PortRowAction.Start("/w/app"), action(DevServerStatus.Idle))
        assertEquals(PortRowAction.Retry("/w/app"), action(DevServerStatus.Failed))
        // A second start while one is coming up would be a second start.
        assertEquals(PortRowAction.OpenSession("s1"), action(DevServerStatus.Starting))
        // Where it is running, and where the interrupt stops it. There is no stop verb on the wire.
        assertEquals(PortRowAction.OpenSession("s1"), action(DevServerStatus.Ready))
        // The protocol says both states carry a session. Drawing a control that would have nowhere
        // to go is worse than drawing none, so the impossible case is handled rather than forced.
        assertEquals(PortRowAction.None, action(DevServerStatus.Starting, session = null))
        assertEquals(PortRowAction.None, action(DevServerStatus.Unknown))

        val plain = LocalhostRow(LocalPort(2019, "wslrelay"), null, null, PortCategory.Other)
        // Nothing on the wire can start or stop "whatever is on 2019".
        assertEquals(PortRowAction.CopyAddress(2019), PortCatalog.secondAction(plain))
    }

    @Test
    fun `the product's two spellings are the ones the brand module produces`() {
        // A rename that reached this client without reaching this file fails here rather than
        // quietly stopping the app's own port from being recognised.
        assertEquals("Terminal Deck", PortCatalog.BRAND_NAME)
        assertEquals("terminaldeck", PortCatalog.BRAND_ID)
    }
}
